closed/adds/jdk/src/share/classes/sun/security/ec/NativeECDHKeyAgreement.java

```diff
@@ -0,0 +1,268 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * ===========================================================================
+ * (c) Copyright IBM Corp. 2022, 2022 All Rights Reserved
+ * ===========================================================================
+ */
+
+package sun.security.ec;
+
+import java.lang.reflect.Method;
+import java.security.AlgorithmParameters;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.Key;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.ProviderException;
+import java.security.PublicKey;
+import java.security.SecureRandom;
+import java.security.interfaces.ECKey;
+import java.security.spec.AlgorithmParameterSpec;
+import java.security.spec.ECGenParameterSpec;
+import java.security.spec.ECParameterSpec;
+import java.security.spec.InvalidParameterSpecException;
+import java.util.HashMap;
+
+import javax.crypto.KeyAgreementSpi;
+import javax.crypto.SecretKey;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.SecretKeySpec;
+
+import jdk.crypto.jniprovider.NativeCrypto;
+
+import sun.security.action.GetPropertyAction;
+import sun.security.util.NamedCurve;
+
+/**
+* Native KeyAgreement implementation for ECDH.
+*/
+public final class NativeECDHKeyAgreement extends KeyAgreementSpi {
+
+    private static final NativeCrypto nativeCrypto = NativeCrypto.getNativeCrypto();
+    private static final String nativeCryptTrace = GetPropertyAction.privilegedGetProperty("jdk.nativeCryptoTrace");
+
+    /* true if OPENSSL_NO_EC2M is defined */
+    private static final boolean noGF2m = nativeCrypto.ECNoGF2m();
+
+    /* stores whether a curve is supported by OpenSSL (true) or not (false) */
+    private static final HashMap<String, Boolean> curveSupported = new HashMap<>();
+
+    /* private key, if initialized */
+    private ECPrivateKeyImpl privateKey;
+
+    /* public key, non-null between doPhase() & generateSecret() only */
+    private ECPublicKeyImpl publicKey;
+
+    /* the type of EC curve */
+    private String curve;
+
+    /* length of the secret to be derived */
+    private int secretLen;
+
+    /* true if noGF2m and ECFieldF2m is used */
+    private boolean useJavaImplementation;
+
+    /* the java implementation, initialized if needed */
+    private ECDHKeyAgreement javaImplementation;
```

Comment:
Why do we need both of these fields? Couldn't `useJavaImplementation` be replaced by `javaImplementation != null`?